### PR TITLE
Take block data for Global Styles from block.json

### PR DESCRIPTION
**Pull request by oandregal**
Follow-up to https://github.com/WordPress/gutenberg/pull/20290 and https://github.com/WordPress/gutenberg/pull/22491

This PR takes the block data necessary for Global Styles (block selector + the supported properties of a block) from the block registry.

## How to test

**Make sure that it works as before:**

- Take this PR.
- Download and activate this [demo theme](https://github.com/nosolosw/global-styles-theme).
- Go to the front end and check that the embedded stylesheet with id `#global-styles-inline-css` has -among other things- these selectors:

```css
h1 {
  font-size: calc(1px * var(--wp--preset--font-size--huge));
  color: var(--wp--preset--color--primary);
}
h4 {
  font-size: var(--wp--preset--font-size--normal);
  color: var(--wp--preset--color--secondary);
}
p {
  line-height: 1.4;
  font-size: calc(1px * var(--wp--preset--font-size--normal));
  color: var(--wp--preset--color--quaternary);
}
```

**Make sure changes in block.json affect the global styles**

- Modify the block.json of the paragraph block. For example, set `__experimentalColor` to false.
- Compile, so block.json is moved to the proper place: `npm run build`.
- Reload the front-end and verify that the paragraph style rule doesn't contain the color declaration.

**Make sure that blocks can only output the properties they support at the block level** 

Some examples to try:

- Add the text color within the global context in the config for the theme. Verify that the generated CSS does not contain a declaration for the color.
- Add a gradient config for the paragraph context in the config for the theme. Verify that the generated CSS does not contain a declaration for the background.

### CODE REVIEW

File: lib/global-styles.php
Comment (youknowriad) [inline lib/global-styles.php:273]: not sure how important is it for this PR but noting that some of these features are not booleans.

For example for the "color" one, you can enable "colors" and "gradients" if  the support for these is enabled, theme.json should be able to define three things:

```
background: gradient;
backgroundColor: red;
color: green;
```

**Comment by gziolo** [comment]
I just wanted to raise that by using a top-level field in the `block.json` brings a few follow-up tasks:
- it has to be included in `WP_Block_Type object on the server
- it needs to be whitelisted as a value that is exposed from the server
- it needs to be added to the schema in the newly created REST API endpoint for block types
- it deserves its own section in the RFC for block registration API
- it deserves its own section in the documentation for the block type shape

I probably missed something :)

File: packages/block-library/src/columns/block.json
Comment (youknowriad) [inline packages/block-library/src/columns/block.json:4]: This should be the default and only needed for blocks that don't have the generated class.

Comment (nosolosw) [reply packages/block-library/src/columns/block.json:4]: ah, so would you prefer doing:

- `selector: true`, for blocks that use the generated class
- `selector: <some-other-selector>` for blocks that don't

Comment (nosolosw) [reply packages/block-library/src/columns/block.json:4]: After [Greg's comment](https://github.com/WordPress/gutenberg/pull/22698#issuecomment-635268128) I'm thinking of moving this under the supports key and do the same as other experimental features, so we'd have:

```
supports: {
  __experimentalSelector: {boolean|string|object}
}
```

File: lib/global-styles.php
Comment (nosolosw) [reply lib/global-styles.php:273]: Yeah, I've been reviewing [the existing style properties](https://github.com/WordPress/gutenberg/issues/22700) and also noticed that background (solid or gradients) was missing from the original hard-coded list in global styles. I'll fix it in this PR.

File: packages/block-library/src/columns/block.json
Comment (youknowriad) [reply packages/block-library/src/columns/block.json:4]: > selector: true, for blocks that use the generated class

Why do we need the `selector: true`. Can't we just consider it as the default if the "selector" is not specified?

Comment (spacedmonkey) [reply packages/block-library/src/columns/block.json:4]: This is not part of the block type RCF is it? @gziolo ?

**Comment by spacedmonkey** [comment]
Should this new field be added to the block type api?

File: lib/global-styles.php
Comment (spacedmonkey) [inline lib/global-styles.php:272]: Don't all these supports need to be added to the json files?

**Comment by jorgefilipecosta** [comment]
Currently, we are using selectors like “core/heading/h1” for global styles. Which means that blocks will need away in the client to know if they are currently affected by a given selector (so their UI can show default values set in theme.json e.g.: font line, height, for a heading 1 etc):
One way to match a block instance to a selector would be that blocks provide matchers something like:
```
{ ‘core/heading/h1’: function( attributes ) => attributes.level === 1, ‘core/heading/h2’: function( attributes ) => attributes.level ===  2 }
```
But in the future, we may also want matchers on the server, e.g: to output CSS only for selectors currently used by some block. And that would require a matcher on the server as well.
What when specifying selectors like core/heading/h1 blocks allow pass the matcher with a syntax like `[level= “1"]` (similar to the native CSS syntax) to target a block only in specific cases?

Regarding the sector specification:
	- Blocks may not set any sector. In that case, it is assumed a block will be targeted by their class.
	- Blocks may set a normal block selector with a custom dom selector `'core/paragraph': 'p'`.
	- Blocks may set multiple selectors and their matches: 'core/heading/h1': [ 'h1', '[level= “1"]'.

File: lib/global-styles.php
Comment (youknowriad) [reply lib/global-styles.php:274]: Noting also that priority is important here (we can't apply the backgroundColor before the gradient otherwise it's overridden)

File: packages/block-library/src/columns/block.json
Comment (jorgefilipecosta) [reply packages/block-library/src/columns/block.json:4]: Yes I agree if no selector is specified at all we should target the block by the default class.

File: packages/block-library/src/heading/block.json
Comment (jorgefilipecosta) [inline packages/block-library/src/heading/block.json:8]: As referred in https://github.com/WordPress/gutenberg/pull/22698#issuecomment-635301127, we will need some way to know if a given block instance is h4 or not.

File: packages/block-library/src/columns/block.json
Comment (jorgefilipecosta) [reply packages/block-library/src/columns/block.json:4]: As automatic selectors we should also use the block styles:
e.g: 'core/quote/large' automatically mapped to .wp-block-quote.is-style-large.

File: lib/global-styles.php
Comment (nosolosw) [reply lib/global-styles.php:272]: This PR proposes to take them from the block.json, where they're already declared. However, this wasn't possible until https://github.com/WordPress/gutenberg/pull/22491 landed. So I guess the answer is: yes!

**Comment by nosolosw** [comment]
@jorgefilipecosta that's a neat idea!

To make sure I fully follow, there are few things to unpack.

- How we target those selectors in `theme.json`. Currently, we use `core/heading/h1`, do you propose we should do `core/heading[level=1]` instead?

- How do we declare those in `block.json`. Would you mind expanding a bit on how would you do this, perhaps with examples for the paragraph and heading blocks?

**Comment by nosolosw** [comment]
> Should this new field be added to the block type api?

@spacedmonkey I think we want to move this under the `supports` key and use the `__experimental` prefix, like the rest of the experimental features. See [thread](https://github.com/WordPress/gutenberg/pull/22698#discussion_r431766247).

I saw that there's [a task](https://core.trac.wordpress.org/ticket/48529) to add the missing fields to the block type API, so I guess whatever we do for the `supports` subkeys that are experimental, we should do for this new field as well.

**Comment by jorgefilipecosta** [comment]
> How we target those selectors in theme.json. Currently, we use core/heading/h1, do you propose we should do core/heading[level=1] instead?


I guess blocks could specify  something like `'core/heading/h1': [ 'h1', '[level= “1"]'`.
But thinking more I think a syntax like [level= “1"] odes does not make much sense we can use JSON directly {level: 1}. And than it seems block variations are already that a set of attributes for the block.
And I guess we can identify if a given block instance matches a specific variation. In that case, maybe we should simply include the ability to specify dom selectors for each block variation?

File: packages/block-library/src/columns/block.json
Comment (jorgefilipecosta) [reply packages/block-library/src/columns/block.json:4]: Yes I agree if no selector is specified at all we should target the block by the default class.

Comment (nosolosw) [reply packages/block-library/src/columns/block.json:4]: > Can't we just consider it as the default if the "selector" is not specified?

Yeah, I think we can.

My mind-frame in coming to this was that it'd be an opt-in mechanism, but it can perhaps work fine by generating the class if no selector is provided. Thinking out loud: what would be the worst that can happen if this isn't opt-in? This is what I can think of:

1. We're duplicating the class generation logic client & server-side. I think this should be fine because that's an area that I don't expect to change.

2. Class conflicts due to namespacing not matching the class name. Ex:
  - `core/media-text` block name => `.wp-block-media-text` class name.
  - `custom-library/media-text` block name => `.wp-block-media-text` class name.

I don't think third-party blocks are overriding core's class names, otherwise, this is problematic on its own. So, perhaps what we should do is to generate this class by default: `.wp-{namespace}-{block-name}` to prevent conflicts. With the exception that we don't add the namespace for core blocks, so they'll be `.wp-{block-name}` instead.

File: lib/global-styles.php
Comment (nosolosw) [reply lib/global-styles.php:274]: I'm thinking that priority can work like it works today (and in CSS, so people's mental models are aligned): all declarations are transformed to CSS in the order they appear in the theme.json. You can't have both a solid & a gradient background, so I think this assumption should be fine.

File: lib/utils.php
Comment (nosolosw) [inline lib/utils.php:2]: This has been ported from https://github.com/WordPress/gutenberg/pull/22744 Will rebase this PR when that lands.

File: lib/global-styles.php
Comment (nosolosw) [inline lib/global-styles.php:84]: If `theme.json` contains a single comma in the wrong place makes `json_decode` to output a void array. As the file grows, this is more likely to happen and more difficult to spot. Adding this code at least helps in quickly determining what's happening by inspecting the error log.

File: lib/global-styles.php
Comment (nosolosw) [inline lib/global-styles.php:305]: This one-liner allows the global context to generate CSS for background-color. Adding it in this PR (which is a refactor to take data from the block.json) because it's so small a change.

File: packages/block-library/src/columns/block.json
Comment (nosolosw) [reply packages/block-library/src/columns/block.json]: > As automatic selectors we should also use the block styles:
e.g: 'core/quote/large' automatically mapped to .wp-block-quote.is-style-large.

@jorgefilipecosta I think adding support for style variations merits its own PR / discussion. The first step would be to look at them as "sets of style attributes" as suggested at https://github.com/WordPress/gutenberg/issues/20331Perhaps after that experiment we end up doing some things (like inlining those values) that remove the need of targeting the existing classes.

**Comment by nosolosw** [comment]
The last bit for this PR is figuring out how the block can determine which selector it uses, which was brought up by @jorgefilipecosta I've done some unsuccessful explorations with [block variations](https://github.com/WordPress/gutenberg/pull/22805) and now I'm looking at attribute sources.

However, I'd think this is PR is ready to be merged and we can iterate on that part in a follow-up PR, in the context in which that part is needed. By merging this I'll be able to unblock some work I plan to do this week, which is to revive [the PR to enable the global styles sidebar in edit-site](https://github.com/WordPress/gutenberg/pull/20530).

So, if @jorgefilipecosta or anyone wants to give this a try and perhaps :heavy_check_mark: if it's ready, I'd happy to merge this as soon as possible.

**Comment by nosolosw** [comment]
This is what I'm thinking for "dynamic selectors" (selectors for blocks that represent many HTML elements, such as core/heading -h1 to h6- and core/list -ol, ul-).

They could be captured like this:

- block.json for core/heading:

```
{
    "name": "core/heading",
    "supports": {
        "__experimentalSelector": {
            "core/heading/h1": {
                "selector": "h1",
                "attributes": {
                    "level": 1
                },
            },
            "core/heading/h2": {
                "selector": "h2",
                "attributes": {
                    "level": 2
                },
            },
        },
    },
}
```

block.json for core/list

```
{
    "name": "core/list",
    "supports": {
        "__experimentalSelector": {
            "core/list/ol": {
                "selector": "ol",
                "attributes": {
                    "ordered": true
                },
            },
            "core/heading/ul": {
                "selector": "ul",
                "attributes": {
                    "ordered": false
                },
            },
        },
    },
}
```

I'm thinking this could work and avoids having a JavaScript callable/function within block.json, so it's parseable by any means we want (JavaScript, PHP, mobile).

If folks are convinced about this approach, I can implement it. However, given that this PR doesn't have a use case for this yet (it's a refactor from hardcoding data to take them from block.json) I think I'll still want to merge as it is and explore this approach where it makes sense (for example, in the edit-site/global styles PR I plan to work on next).

**Comment by jorgefilipecosta** [comment]
Hi @nosolosw,
```
              "core/heading/h1": {
                "selector": "h1",
                "attributes": {
                    "level": 1
                },
            },
```


This approach makes sense to me. My only worry is that we are just specifying a set of attribute values. What is the difference between block variations? Conceptually block variations are just a set of attribute values. To see if the block matches core/heading/h1 we will need to see if the attributes of the block are equal to attributes. Could we just do the same with variations and verify if the attributes of a block match a variation?

**Comment by jorgefilipecosta** [comment]
There is a also a possible solution that would not involve specifying the set of attributes. It is using dom functions and verifying if the block dom container matches the selector using the matches function https://developer.mozilla.org/en-US/docs/Web/API/Element/matches.
This solution involves accessing dom then using effects to apply the styles or not to the context, it has some performance issues so I'm not sure it is something we should try.

File: lib/global-styles.php
Comment (jorgefilipecosta) [inline lib/global-styles.php:84]: Comparing by the error message does not seem very reliable. The error messages may change, or localization may impact it. I think we should use https://www.php.net/manual/en/function.json-last-error.php.

File: lib/global-styles.php
Comment (jorgefilipecosta) [inline lib/global-styles.php:359]: We are using a regex replace with preg_replace which is more costly that a simple str_replace. I think in this case we are not taking advantage of a regex so str_replace should be enough.

**Review by jorgefilipecosta — APPROVED** [review]
I think we can merge this PR so we can continue the iterations.
There are some followups that should be proposed after this PR is merged:
- Have a way to match a selector to a given block.
- Automatically generate selectors based on block styles. So block styles can be defined with theme.json.

File: lib/global-styles.php
Comment (nosolosw) [reply lib/global-styles.php]: Changed it to `str_replace` at https://github.com/WordPress/gutenberg/pull/22698/commits/b2a77df59d3fda32f84a8072e6ec09c660f4326f

File: lib/global-styles.php
Comment (nosolosw) [reply lib/global-styles.php]: Changed it to check the error code at https://github.com/WordPress/gutenberg/pull/22698/commits/3346f246ed2bf65c669961249464c6c5e3df5c87

**Comment by nosolosw** [comment]
Rebased to fix a legit CI error that was presumably fixed by [this](https://github.com/WordPress/gutenberg/pull/22785).

**Pull request merged**
